File: 11-Binary-Search/cpp/02-matrix-median.cpp
class Solution {
public:
    int findMedian(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();

        int low = 1, high = 1e9;
        int required = (n * m) / 2;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            int count = countSmallerOrEqual(matrix, mid);

            if (count <= required) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return low;
    }

private:
    int countSmallerOrEqual(const vector<vector<int>>& matrix, int x) {
        int count = 0;
        for (const auto& row : matrix) {
            count += upper_bound(row.begin(), row.end(), x) - row.begin();
        }
        return count;
    }
};

File: 11-Binary-Search/python/02_matrix_median.py
import bisect

class Solution:
    def findMedian(self, matrix: List[List[int]]) -> int:
        n, m = len(matrix), len(matrix[0])
        low, high = 1, 10**9
        required = (n * m) // 2

        def count_smaller_or_equal(x):
            return sum(bisect.bisect_right(row, x) for row in matrix)

        while low <= high:
            mid = (low + high) // 2
            if count_smaller_or_equal(mid) <= required:
                low = mid + 1
            else:
                high = mid - 1

        return low
